weejson/src/main/scala/com/rallyhealth/weejson/v1/BufferedValue.scala

```diff
@@ -172,15 +186,57 @@ object BufferedValue extends Transformer[BufferedValue] {
   sealed trait AnyNum extends BufferedValue {
     def value: BigDecimal
   }
+
   case class Num(s: String, decIndex: Int, expIndex: Int) extends AnyNum {
-    override def value: BigDecimal = BigDecimal(s)
+    override lazy val value: BigDecimal = BigDecimal(s)
+
+    override def equals(that: Any): Boolean = that match {
+      case NumLong(l) => value == l.value
+      case NumDouble(d) =>
+        val thisD = value.toDouble // may chop precision or go infinite
+        if (thisD.isInfinite) value == d.value else thisD == d
+      case other: Num => value == other.value
+      case _ => super.equals(that)
+    }
+
+    override def hashCode(): Int = {
+      /*
+       * All values outside of Double's range hash to the same values
+       * (i.e., Double.PositiveInfinity.## and Double.NegativeInfinity.##),
+       * but these gigantic numbers are rarely encountered.
+       */
+      this.value.toDouble.##
+    }
   }
+
   case class NumLong(l: Long) extends AnyNum {
     override def value: BigDecimal = BigDecimal(l)
+
+    override def equals(that: Any): Boolean = that match {
+      case NumLong(otherL) => this.l == otherL
+      case NumDouble(otherD) => this.l.toDouble == otherD
```

Comment:
Does `l == otherD` work here? Tests seem to pass with it.

Reply:
You are right, I'll fix it. I was trying to avoid the subtleties with |L| > 2^53, but it looks like `Long.equals` already accommodates that. However, the inverse is not true, i.e., `this.l == otherD.toLong` will not work for all values outside of the 2^53 range.